torch/fx/experimental/symbolic_shapes.py

```diff
@@ -3627,7 +3627,7 @@ def hint(s):
 
             for src, curr_t, constraint in sources_tensors_constraints:
                 if is_sparse_any(curr_t):
-                    for i, ss in enumerate(curr_t.size()):
+                    for i, ss in enumerate(curr_t.size()[:curr_t.sparse_dim()]):
```

Comment:
Don't we still have to generate sizes for the dense dims?

Reply:
Right, the fix must be done in the other end (to be determined) that appears to ignore dense dimensions.

Reply:
The failing test has
```
        r.sparse_resize_((4, 4), 1, 1)
        r.sparse_resize_and_clear_((4, 4, 4), 2, 1)
```
that is, the test changes the dimensionality of the tensor in-place which will confuse dynamo..

Reply:
A minimal reproducer:
```python
    @parametrize("dtype", [torch.float64])
    def test_meta_coo_tensor_(self, dtype):
        r = torch.sparse_coo_tensor(size=(4, 4), device='meta', dtype=dtype)
        r.sparse_resize_and_clear_((4, 4, 4), 2, 1)
```

Traceback:

<details>

```
Traceback (most recent call last):
  File "/home/pearu/git/aartbik/pytorch/test/test_sparse.py", line 4258, in test_meta_coo_tensor_1
    r = torch.sparse_coo_tensor(size=(4, 4), device='meta', dtype=dtype)
  File "/home/pearu/git/aartbik/pytorch/torch/_dynamo/convert_frame.py", line 986, in catch_errors
    return callback(frame, cache_entry, hooks, frame_state, skip=1)
  File "/home/pearu/git/aartbik/pytorch/torch/_dynamo/convert_frame.py", line 827, in _convert_frame
    result = inner_convert(
  File "/home/pearu/git/aartbik/pytorch/torch/_dynamo/convert_frame.py", line 410, in _convert_frame_assert
    return _compile(
  File "/home/pearu/git/aartbik/pytorch/torch/_utils_internal.py", line 70, in wrapper_function
    return function(*args, **kwargs)
  File "/home/pearu/miniconda3/envs/pytorch-cuda-dev/lib/python3.9/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/home/pearu/git/aartbik/pytorch/torch/_dynamo/convert_frame.py", line 737, in _compile
    raise InternalTorchDynamoError(str(e)).with_traceback(
  File "/home/pearu/git/aartbik/pytorch/torch/_dynamo/convert_frame.py", line 708, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
  File "/home/pearu/git/aartbik/pytorch/torch/_dynamo/utils.py", line 273, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/pearu/git/aartbik/pytorch/torch/_dynamo/convert_frame.py", line 665, in compile_inner
    check_fn = CheckFunctionManager(
  File "/home/pearu/git/aartbik/pytorch/torch/_dynamo/guards.py", line 2086, in __init__
    guard.create(builder)
  File "/home/pearu/git/aartbik/pytorch/torch/_guards.py", line 259, in create
    return self.create_fn(builder, self)
  File "/home/pearu/git/aartbik/pytorch/torch/_dynamo/guards.py", line 1683, in SHAPE_ENV
    guards = output_graph.shape_env.produce_guards(
  File "/home/pearu/git/aartbik/pytorch/torch/fx/experimental/symbolic_shapes.py", line 3636, in produce_guards
    track_symint(property_source, ss, constraint[i])
torch._dynamo.exc.InternalTorchDynamoError: list index out of range
```

</details>

However, the following tests work fine:
```python
    @parametrize("dtype", [torch.float64])
    def test_meta_coo_empty(self, dtype):
        r = torch.empty(4, 4, layout=torch.sparse_coo, device='meta', dtype=dtype)
        r.sparse_resize_and_clear_((4, 4, 4), 2, 1)
```
or
```python
    @parametrize("dtype", [torch.float64])
    def test_meta_coo_tensor_2(self, dtype):
        r = torch.sparse_coo_tensor(size=(4, 4), device='meta', dtype=dtype)
        r.sparse_resize_and_clear_((4, 4, 4), 2, 1)
        # print(r)        # test fails when enabling this print
        print(f"{r=}")  # test passes
```

Reply:
The case with `print(f"{r=}")` passes only because `repr` on TensorVariable is unsupported and the `transform` is force to finish earlier, before `InternalTorchDynamoError` is raised.

Reply:
OK, here's a fix:
```
$ git diff
diff --git a/torch/_dynamo/variables/tensor.py b/torch/_dynamo/variables/tensor.py
index e928a9e0ea6..f951480c2c1 100644
--- a/torch/_dynamo/variables/tensor.py
+++ b/torch/_dynamo/variables/tensor.py
@@ -745,6 +745,12 @@ class TensorVariable(VariableTracker):
     def method_resize_as_(self, *args, **kwargs):
         unimplemented("Tensor.resize_as_")
 
+    def method_sparse_resize_(self, *args, **kwargs):
+        unimplemented("Tensor.sparse_resize_")
+
+    def method_sparse_resize_and_clear_(self, *args, **kwargs):
+        unimplemented("Tensor.sparse_resize_and_clear_")
+
     def method_set_(self, *args, **kwargs):
         if len(args) > 1:
             # torch.Tensor.set_() has several overloads.
diff --git a/torch/fx/experimental/symbolic_shapes.py b/torch/fx/experimental/symbolic_shapes.py
index 245d19e48c8..9b6c97f90a9 100644
--- a/torch/fx/experimental/symbolic_shapes.py
+++ b/torch/fx/experimental/symbolic_shapes.py
@@ -3627,7 +3627,7 @@ class ShapeEnv:
 
             for src, curr_t, constraint in sources_tensors_constraints:
                 if is_sparse_any(curr_t):
-                    for i, ss in enumerate(curr_t.size()[:curr_t.sparse_dim()]):
+                    for i, ss in enumerate(curr_t.size()):
                         property_source = TensorPropertySource(src, TensorProperty.SIZE, i)
                         track_symint(property_source, ss, constraint[i])
                 else:
```

Reply:
graph breaking sgtm

Reply:
do you want me to apply the fix above into this PR, or is that for a follow up?

Reply:
> do you want me to apply the fix above into this PR, or is that for a follow up?

In this PR because my previous suggested change to torch/fx/experimental/symbolic_shapes.py was wrong.